tests/unit/test_charm.py

```diff
@@ -1,105 +1,138 @@
 # Copyright 2021 Canonical Ltd.
 # See LICENSE file for licensing details.
 
-import json
-
 import pytest
 import yaml
-from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+from ops import BlockedStatus
+from ops.model import ActiveStatus
 from ops.testing import Harness
 
-from charm import Operator
+from charm import EnvoyOperator
 
 
 @pytest.fixture
-def harness():
-    return Harness(Operator)
+def harness(mocked_kubernetes_service_patch):
+    """Harness populated with the Charm and with its KubernetesServicePatch patched."""
+    return Harness(EnvoyOperator)
 
 
-def test_not_leader(harness):
-    harness.begin_with_initial_hooks()
-    assert harness.charm.model.unit.status == WaitingStatus("Waiting for leadership")
+@pytest.fixture()
+def mocked_kubernetes_service_patch(mocker):
+    """Mocks the KubernetesServicePatch for the charm."""
+    mocked_kubernetes_service_patch = mocker.patch(
+        "charm.KubernetesServicePatch", lambda x, y: None
+    )
+    yield mocked_kubernetes_service_patch
 
 
-def test_missing_image(harness):
-    harness.set_leader(True)
-    harness.begin_with_initial_hooks()
-    assert harness.charm.model.unit.status == BlockedStatus("Missing resource: oci-image")
+class TestCharm:
 
+    def test_not_leader(self, harness):
+        """Test that the charm is not active when not leader."""
+        harness.begin_with_initial_hooks()
+        assert "Waiting for leadership" in harness.charm.leadership_gate.status.message
+        assert not isinstance(harness.charm.model.unit.status, ActiveStatus)
 
-def test_no_relation(harness):
-    harness.set_leader(True)
-    add_oci_image(harness)
+    def test_no_grpc_relation(self, harness):
+        """Test the grpc Component and charm are not active when no grpc relation is present."""
+        harness.set_leader(True)
+        harness.begin_with_initial_hooks()
 
-    harness.begin_with_initial_hooks()
+        assert (
+            "Expected data from exactly 1 related applications - got 0"
+            in harness.charm.grpc.status.message
+        )
+        assert not isinstance(harness.charm.model.unit.status, ActiveStatus)
```

Comment:
The thought behind this was that if this component has a Blocked status then:
* the charm cannot be active (that's something the `CharmReconciler` enforces - if any `Component` is blocked, then the `CharmReconciler` puts the charm to blocked)
* the charm might be blocked by **this** component, but that is not guaranteed because some other `Component` could have blocked the charm first (for example, if we make the ingress relation required)

So I think asserting the charm is blocked by **this** component adds a flaky element to the test.  wdyt?

Thinking on this more, do we need any assertion on the `unit.status`?  If we've asserted above `harness.charm.grpc.status.message` has the error message, the only way this doesn't get to the charm is if there's a `CharmReconciler` bug.  So this assertion on `unit.status` is I think just testing the `CharmReconciler`.

Reply:
For now, I've added some assertions to be certain the `Component` is in blocked (before we just checked the message).  But we can do more changes if you like

Reply:
Ok, added a logger to the charm and some tests.  wdyt?

Reply:
> Thinking on this more, do we need any assertion on the unit.status? If we've asserted above harness.charm.grpc.status.message has the error message, the only way this doesn't get to the charm is if there's a CharmReconciler bug. So this assertion on unit.status is I think just testing the CharmReconciler.

The assertion of the message seems right.

About how the `CharmReconciler` handles errors and unit status, I think we can have a discussion later.